src/librustc_mir/interpret/const_eval.rs

```diff
@@ -560,12 +562,31 @@ pub fn const_eval_provider<'a, 'tcx>(
 
     let (res, ecx) = eval_body_and_ecx(tcx, cid, None, key.param_env);
     res.and_then(|(mut val, ptr, layout)| {
-        ecx.validate_ptr_target(
-            ptr.to_ptr()?,
-            layout.align,
-            layout,
-            String::new(),
-        )?;
+        let ptr = ptr.to_ptr()?;
+        let mut todo = vec![(ptr, layout.ty, String::new())];
+        let mut seen = FxHashSet();
+        seen.insert((ptr, layout.ty));
+        while let Some((ptr, ty, path)) = todo.pop() {
+            if let Some(AllocType::Static(did)) = tcx.alloc_map.lock().get(ptr.alloc_id) {
+                if did == def_id {
+                    // prevent recursive statics from trying to self-validate
+                    // this would be nice-to-have, but we are inside the query that
+                    // evaluates said static, and const eval will try to run this query again
+                    // if we do this kind of validation
+                    // We could run the checks as an additional lint after evaluation is done
```

Comment:
Now I'm thinking we should totally do this generally. If we accidentally make any checks too strict, users can opt out of them for specific constants until we fix it.

Reply:
What do you mean by "generally"?

Reply:
Adding new checks to const eval as lints instead of breaking the evaluation

Reply:
Deny-by-default lint? Yeah I guess that makes sense. We *can* continue compilation, after all. It just probably doesn't do what people want (as in, it raises UB...).